### ISSUE TITLE
Support for buffered writes

### ISSUE THREAD

**Issue by 256dpi** [gorilla/websocket#139]
First of all thanks for the amazing quality of this library!

I'm working on gomqtt (an implementation of the MQTT protocol) that provides multiple packages to abstract the protocol. One of the packages is [transport](https://github.com/gomqtt/transport) that provides an interface for TCP and WebSocket connections. For the later I'm obviously using the `gorilla/websocket` package.

Currently I'm focusing on reducing syscalls by buffering outgoing data and flush the buffers asynchronously: https://github.com/gomqtt/transport/blob/buffered-send/net_conn.go#L96. That way I was able to increase the throughput of TCP connections 10 times.

Unfortunately the `gorilla/websocket` package provides currently no interface for managing the Writes to the underlying connection. I know that the package is using the concept of a `MessageWriter` and therefore adding that low level management would probably interfere with this nice abstraction. Yet, I'm wondering if you see any possibility that something could be implemented that allows managing the writes and flushes to the underlying connection?

I'm happy to contribute in this matter, if you could point me in the right direction.

In the IoT world, devices tend to send many small messages, and as protocols like MQTT also offer a WebSocket based solution, such a feature might become more important in the future.

Looking forward to get some feedback!

**Comment by garyburd** [comment]
The package buffers output within a message and flushes at the end of each message. 

What specifically do you want to change?

**Comment by 256dpi** [comment]
A typical MQTT broker might have a throughput of several hundred thousand messages per second. Sending a message over WebSocket currently generates one syscall per message. In the mentioned performance improvement, the TCP layer combines multiple messages into one syscall.

Since the MQTT over WebSocket spec defines that one MQTT packet is sent using one WebSocket message the said improvement cannot be implemented.In that sense I would like to be able to write multiple WebSocket messages in a buffer and periodically flush that buffer to the underlying connection.

I hope its more clear now.

**Comment by garyburd** [comment]
It's clear now. An application can do this today in a wrapper around the underlying network connection. 

Why would an application use MQTT over Websocket when performance is important?

**Comment by 256dpi** [comment]
Wrapping the underlying connection would be a solution. But the only interface that would currently allow something like that is the `Dialer.NetDial` callback. The `Upgrader` has no such callback.

In any case, using such callbacks to wrap the connection seems a little to hacky for me. It would be cool to have some additional functions on the connection that provide more fine grained control like:

```
BufferMessage(messageType int, data []byte) error
FlushMessages() error
```

To support these methods the `NextWriter` method could additionally support a variadic parameter flush that indicates if the buffer should be flushed after closing the writer.

```
NextWriter(messageType int, flush... bool) (io.WriteCloser, error)
```

Of course MQTT over plain TCP is the fastest and most used method. All tough, MQTT gets more and more adopted by the web community. In most cases such a website would receive data from multiple sensors and display it using some charts. While the receiving part does not need to be that performant, I don't want to have my broker wait on syscalls when forwarding these messages to the web socket connection.

Do you think such an use case should be a concern of this library?

**Comment by garyburd** [comment]
An application can wrap the underlying connection on a server by implementing the [Hijacker](https://godoc.org/net/http#Hijacker) interface.

See the [chat application](https://github.com/gorilla/websocket/tree/master/examples/chat) for an example of a broker that does not wait on syscalls when forwarding messages to a websocket connection.

I'll comment more next week.

**Comment by garyburd** [comment]
The [PR](https://github.com/gorilla/websocket/issues/139) adds a second level of buffering. The cost of the second level of buffering provides no benefit to most users of the package.

I am not convinced the use case should be a concern for the package.  Is there a reason why an application will use MQTT over WebSockets when performance is a important?

**Comment by 256dpi** [comment]
Using two buffers is for shure redundant, in all cases it would be safe to use the new writer rather than `writeBuf`. I'll update the PR accordingly this week.

An example is https://shiftr.io, the project I'm currently working on. The service simplifies the usage of a MQTT broker for makers, hobbyists and students. A big feature is the graph visualization of a namespace that shows all connections, subscriptions and messages in real-time. Thus, the broker has to deal with many firehose connections over the WebSocket protocol. Having a way to control (buffer) the writes to these connections would save lots of resources.

As this library is imposing a certain interface to use the WebSocket protocol and does not provide `Encode()` or `Decode()` functions that would allow to implement a custom connection, this use case must be a concern fore the package. Otherwise a user like me has to mantain a fork of the package.

**Comment by garyburd** [comment]
I suggest that we agree on the problem to be solved before you write more code.

Is MQTT over websockets the use case that we are talking about?  Are you using MQTT over websockets for shiftr.io?

What are the Encode() and Decode() functions?

An application can implement buffering similar to that in the PR with the package as is.

**Comment by 256dpi** [comment]
Yes, the real-time visualization is based on a MQTT-over-WebSocket connection that makes a firehose-like subscription to receive all published messages and other system related data. While embedded devices of course us a plain MQTT connection, there are also users that use the browser to visualize data. So, MQTT-over-WebSocket is not rarity in my case and is used alongside normal MQTT.

With Encode() and Decode() functions I think of functions that in other words serialize WebSocket frames to a buffer or deserialize them from a buffer. This package has currently no such public functions that would allow to be used to implement a custom Conn type for example.

I currently do not see a simple and clean way to implement that kind of buffering, do you have an example?

**Comment by garyburd** [comment]
Why does shiftr.io use MQTT over websocket instead of normal MQTT for real-time visulatization?

**Comment by 256dpi** [comment]
In my dimension browsers do not support raw TCP sockets.... do they in yours? ;)

Sorry, I missed your point here.

**Comment by garyburd** [comment]
I asked the question because I did not understand your use case. Is this your use case:

The server and browser client are communicating with MQTT over websockets. The browser client is using an MQTT client written in Javascript. The server sends messages to a single client at a rate high enough that there's benefit to coalescing messages at both the syscall level and on the network.

I want to understand the the use case and its importance before adding features to support it. New features can make the package more complex to use and add to my support costs. With the proposed API, users of the package will need to decide if they should use BufferMessage or WriteMessage. Some of the users who pick BufferMessage will file an issue complaining about latency.

Here's a rough draft of adding buffering on the server using the API as is:

     type hijacker struct {
          http.ResponseWriter
          bw *bufio.Writer
     }

     func (h *hijacker)  Hijack() (net.Conn, *bufio.ReadWriter, error) {
           c, rw, err := h.ResponseWriter.Hijack()
           if rw != nil {
                h.bw = rw.Writer
                c = &conn{c, rw.Writer}
           }
           return c, rw, err
    }

     type conn struct {
         net.Conn
         bw *bufio.Writer
    }

    func (c *conn) Write(p []byte) (int, error) {
          return c.bw.Write(p)
    }

    ...

    h := &hijacker{w}
    c, err := upgrader.Upgrade(h, r, nil)
    if err != nil {
         // handle error
    }
    bw := h.bw
    // call bw.Flush() as needed.

**Comment by 256dpi** [comment]
Yes, thats precisely my use case.

I can understand that this API addition adds complexity and raises maintenance costs. But as I have said in earlier comments, my feeling is that this should be part of the library as it covers the missing part to achieve highly performant WebSocket servers. If this would be in the responsibility of the user, you would have to explain the hijacker approach (which has a little hacky touch) to users like me that are not that familiar with that API.

Why do you think users might file an issue about latency? If both methods are properly documented, it should be very transparent what the tradeoff between them is.

**Comment by garyburd** [comment]
The MQTT spec [says](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718127): 

> A single WebSocket data frame can contain multiple or partial MQTT Control Packets. The receiver MUST NOT assume that MQTT Control Packets are aligned on WebSocket frame boundaries

If your JS client supports the spec, then you can achieve your goal with the following code:

    type WebSocketConn struct {
	  conn *websocket.Conn
	  wc   io.WriteCloser
	  flushTimer *time.Timer
	  flushError error
        ...
    }

    func (c *WebSocketConn) Send(pkt packet.Packet) error {
	  c.sMutex.Lock()
	  defer c.sMutex.Unlock()

	  // reset and eventually grow buffer
	  packetLength := pkt.Len()
	  c.sendBuffer.Reset()
	  c.sendBuffer.Grow(packetLength)
	  buf := c.sendBuffer.Bytes()[0:packetLength]

	  // encode packet
	  n, err := pkt.Encode(buf)
	  if err != nil {
		c.end(websocket.CloseInternalServerErr)
		return newTransportError(EncodeError, err)
	  }

	  if c.wc == nil {
		c.wc, err = c.conn.NextWriter(websocket.BinaryMessage)
		if err != nil {
			// handle error
		}
	  }
	  // write packet to connection
	  err = c.wc.Write(buf)
	  if err != nil {
		c.end(websocket.CloseInternalServerErr)
		return newTransportError(NetworkError, err)
	  }

	  // increment write counter
	  atomic.AddInt64(&c.writeCounter, int64(n))

	  // queue asyncFlush
	  c.flushTimer.Reset(flushTimeout)

	  return nil
    }

    func (c *WebSocketConn) asyncFlush() {
	  c.sMutex.Lock()
	  defer c.sMutex.Unlock()

	  // flush buffer and save an eventual error
	  if c.wc != nil {
		err := c.wc.Close()
		c.wc = nil
		if err != nil {
			c.flushError = err
		}
	  }
    }

Another approach is to run a goroutine to pump packets to the websocket connection where the goroutine writes multiple packets to a single websocket message:

    type WebSocketConn struct {
	  conn *websocket.Conn
	  send chan packet.Packet  // queue of packets to send, capacity should be > 0
	  ...
    }

     func (c *WebSocketConn) Send(pkt packet.Packet) error {
         c.send <- pkt
         // add code to return errors detected in writePump.
    }

    func (c *WebSocketConn) writePump() {

        // run this method as goroutine. Close c.send to exit the goroutine.

        for  {
            pkt, ok := <- c.send
            if !ok {
               // writer close send channel. Clean up and return
            }
            w, err := c.ws.NextWriter(websocket.BinaryMessage)
            if err != nil {
                // handle error
            }

            // encode pkt to buf as in previous snippet
            ...
            w.Write(buf)

            // Add queued packets to the current websocket message.
            n := len(c.send)
            for i := 0; i < n; i++ {
                pkt = <-c.send
                // encode pkt to buf as in previous snippet
                ...
                w.Write(buf)
             }
             if err := w.Close(); err != nil {
                  // handle error
            }
        }
    }

With this approach, the caller of WebSocketConn.Send does not block when there's available capacity in the send channel.  The goroutine can also be used to send pings to check the health of the client.  See the [chat application](https://github.com/gorilla/websocket/tree/master/examples/chat) for a complete example of this general approach.

**Comment by 256dpi** [comment]
Thanks for pointing out that section in the spec, I always assumed that a MQTT package is tied to one WebSocket frame. But you're right, the JavaScript client does actually support the spec and therefore it should work.

I successfully updated my package based on the code from your first example and can report that the throughput is now even better due to the usage of less WebSocket messages/frames. I also updated the package to support chunked and coalesced messages, as this part was missing due to the wrong assumption.

I'll close the issue and cancel the pull request as the problem has been solved for me and I can tell from the discussion that you'd like to keep the API as is, until it would become absolutely necessary.

Thanks for your persistency and time to discuss this!
